src/cmd/launch.rs

```diff
@@ -1,15 +1,119 @@
+    NotImplemented,
+    Fork,
+    Waitpid,
+    UnmountFailed(nix::errno::Errno),
+    MountFailed(nix::errno::Errno),
+    InputValue,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            AlreadyExists => write!(f, "Debug Container Already Exists"),
+            Umount => write!(f, "umount error"),
+            NonValidUnicode => write!(f, "non valid unicode"),
+            InvalidRootFSPath => write!(f, "invalid rootfs path"),
+            NotImplemented => write!(f, "Not implemented"),
+            Fork => write!(f, "failed fork"),
+            Waitpid => write!(f, "failed waitpid"),
+            InputValue => write!(f, "Input value is illegal"),
+        }
+    }
+}
+
+impl error::Error for Error {}
 
 pub struct LaunchStruct;
 
 impl<DO> Launch<DO> for LaunchStruct
 where
     DO: Downloader,
 {
+    /// rootfsの種類などが記載された設定ファイルsetting.yamlを~/.injesh/containers/に作成する
+    /// /var/lib/docker/overlay2/<HASH_ID>/upperを~/.injesh/containers/<hoge>/upperに対してコピーする
+    /// デバック対象コンテナのlowerdirに対してrootfsを追加した後reloadする
+    /// デバック対象コンテナのプロセスIDを取得
+    /// forkする
+    /// 取得したデバック対象コンテナプロセスIDをもとにsetnsをし、名前空間を同一にする
+    /// 与えられた初期実行ファイルをexecする
     fn launch(&self, launch: &command::Launch<DO>) -> Result<(), Box<dyn std::error::Error>> {
-        println!("execute launch!");
+        // injeshコマンドが初期化されてるかどうかチェック
+        utils::check_initialized()?;
+
+        // デバックコンテナの設定ファイル、ディレクトリ周りの初期化
+        initialize_setting(launch)?;
+
+        // overlyafsのマウントし直し
+        remount(launch)?;
+
+        // TODO: デバック対象コンテナのlowerdirに対してrootfsを追加した後reloadする
+
+        // デバック対象コンテナのプロセスIDとネームスペースのファイルディスクリプタを取得
+        let container_pid = launch.target_container().pid();
+        let ns = Ns::new(container_pid)?;
+
+        unsafe {
+            match fork() {
+                // 親プロセスの場合
+                Ok(ForkResult::Parent { child, .. }) => match waitpid(child, None) {
+                    Ok(status) => println!("Child exited {:?}", status),
+                    Err(_) => Err(Error::Waitpid)?,
+                },
+                // 子プロセス
+                Ok(ForkResult::Child) => {
+                    // setnsで名前空間を変更
+                    ns.setns_net()?;
+                    ns.setns_cgroup()?;
+                    ns.setns_ipc()?;
+                    ns.setns_pid()?;
+
+                    // これがあるとなぜか失敗する
+                    // setns(user_fd.as_raw_fd(), CloneFlags::empty())?;
+
+                    ns.setns_mnt()?;
+                    ns.setns_uts()?;
+
+                    // execでプログラムを実行
+
+                    let main = CString::new(launch.cmd().main())?;
+
+                    let detail: Vec<CString> = launch
+                        .cmd()
+                        .detail_iter()
+                        // TODO: mapないでErrorが生じた場合どのようにするのがベスト？
+                        .map(|d| CString::new(d).unwrap())
+                        .collect();
```

Comment:
mapだけではエラーハンドリングが難しいので、`filter_map`あたりを使うとよさげに思う

このようにすると`command::Cmd`の`detail_iter`メソッドを除去可能

```suggestion
                    let detail: Vec<CString> = launch
                        .cmd()
                        .detail()
                        .iter()
                        .filter_map(|s| CString::new(s.as_str()).ok())
                        .collect::<Vec<CString>>();
```